### ISSUE TITLE
Listing every format that might conceivably be represented as virtual zarr

### ISSUE THREAD

**Issue by TomNicholas** [zarr-developers/VirtualiZarr#218]
**Let's list all the file formats that could potentially be represented efficiently as "virtual zarr" - i.e. zarr + chunk manifests.**

The important criteria here is that the format must store data in a contiguous chunks, such that access using http range requests to object storage is efficient. This rules out some formats, for example I don't think we can efficiently access this format that @kmuehlbauer mentioned over in https://github.com/openradar/xradar/issues/187#issuecomment-2271327041:

> file formats where variables are written interleaved within one chunk of data (eg: 100 bytes v1, 100 bytes v2, 100 bytes v3, 100 bytes v1, 100 bytes v2, 100 bytes v3, ...)? Is there something like strides available?

If we start thinking of Zarr as a "SuperFormat" (super as in superset, not as in super-duper), then this is the list of existing formats comprising that set of what can be referenced using chunk manifests (see https://github.com/zarr-developers/zarr-specs/issues/287).

---

**Definitely can support:**

- [x] HDF5
- [x] NetCDF4
- [ ] GRIB
- [x] TIFF
- [x] FITS
- [x] DMR++ https://github.com/zarr-developers/VirtualiZarr/issues/85

**Probably can support:**

- [ ] HDF4 https://github.com/zarr-developers/VirtualiZarr/issues/216
- [ ] arbitrary collections of `.npz` files

**Maybe can support?**

- [ ] CSV https://github.com/zarr-developers/VirtualiZarr/issues/200
- [ ] Virtual Rasters https://github.com/zarr-developers/VirtualiZarr/issues/166
- [ ] rainbow https://github.com/openradar/xradar/issues/187#issuecomment-2271327041

**Probably can't support:**

- furuno https://github.com/openradar/xradar/issues/187#issuecomment-2271327041
- sigmet/nexrad https://github.com/openradar/xradar/issues/187#issuecomment-2271327041

(The checkboxes indicate whether or not a working implementation already exists - going through kerchunks' in-memory format as an intermediate or creating a `ManifestArray` directly.)

cc @jhamman @d-v-b

**Comment by martindurant** [comment]
I like this issue! It's worth saying that anything kerchunk can chunk can be v-zarrred, right? In that repo, there are suggestions of other worthwhile formats, dicom and nifti (medical imaging) spring to mind. The latter is nice, but often whole-file-gzipped, the former is evil in the way that other 90s standards are evil, but extremely widespread.

**Comment by norlandrhagen** [comment]
> ... the former is evil in the way that other 90s standards are evil, but extremely widespread.

❤️

**Comment by TomNicholas** [comment]
> anything kerchunk can chunk can be v-zarrred, right?

Yes, that's the idea. [This function](https://github.com/zarr-developers/VirtualiZarr/blob/fdab54cca8d73af4e44903091e5e00bbeb1bb31c/virtualizarr/xarray.py#L314) does `kerchunk refs -> virtual dataset`, and [this function](https://github.com/zarr-developers/VirtualiZarr/blob/fdab54cca8d73af4e44903091e5e00bbeb1bb31c/virtualizarr/kerchunk.py#L220) does `virtual dataset -> kerchunk refs`. Any additional kerchunk file readers can be called as another `if...else...` in [here](https://github.com/zarr-developers/VirtualiZarr/blob/fdab54cca8d73af4e44903091e5e00bbeb1bb31c/virtualizarr/kerchunk.py#L91).

**Comment by TomNicholas** [comment]
Hugging Face [safetensors](https://huggingface.co/docs/safetensors/index) is an interesting example - it's uncompressed so basically just like reading netCDF3, having no internal chunking. But it also puts all the metadata at the start of the file, making it a bit like cloud-optimized HDF5. See also https://github.com/huggingface/safetensors/issues/527#issuecomment-2567148998

**Comment by martindurant** [comment]
If the format is simple and common, I say it should be included immediately, especially when there is a straight-forward way to check correctness.

> having no internal chunking

but you *can* assign internal chunking. Is partial reading available in upstream at all yet?

**Comment by TomNicholas** [comment]
> If the format is simple and common, I say it should be included immediately, especially when there is a straight-forward way to check correctness.

I raised #367 to track adding it.

> but you can assign internal chunking. Is partial reading available in upstream at all yet?

This issue seems to suggest it has: https://github.com/zarr-developers/zarr-python/issues/1106. But I think to take advantage of this with virtualizarr would require #199 to be merged.

**Comment by martindurant** [comment]
No, zarr's PR1106 only implemented it for blosc compression, something I've been arguing about for a very very long time!

If you can dynamically re-imagine the chunking at runtime (which is what I tink #119 does), that that would be good enough for most practical uses - but still annoying. Zarr should just do this! i.e., the chunk IO function shouldn't just be passed "I need chunk X", but "I need section (:, s:t, i:j) of chunk X" and a way to characterise what the decompression pipeline looks like (this is OK for decompressed, some blosc, zstd maybe..., but not zlib). This was my suggestion in passing Contexts around in zarr v2.

**Comment by TomNicholas** [comment]
I don't disagree, but if we want to discuss this further we should do it on a new issue (on this repo or upstream on zarr).

**Comment by itsgifnotjiff** [comment]
Environment and Climate Change Canada here 😊. We have something called RPN/FST files which are binary files containing metadata, grid and fortran arrays. 

They are incredibly efficient in both disk space and compression (optional).

[c/Fortran library to work with them](https://github.com/ECCC-ASTD-MRD/librmn)
[Python library that opens them into xarray Datasets with attrs and all](https://github.com/neishm/fstd2nc)

We would love to use something like VirtualiZarr to create per model data cubes that can be easily accessed by ML modules like Jax/torch/Dask but also by Web APIs such as pygeoapi for OGC APIs. 

Some of the difficulties are the fact that we have multiple grids and vertical dimensions per file and not all files of a model contain the same variables (accumulation variables start after 0 forecast hour).

Indexing our data lake is something doable but up until recent advancements in both Zarr indexing and Grouping it did not really fit our plans.

Tools like [SHTools](https://shtools.github.io/SHTOOLS/), [XPublish](https://xpublish.readthedocs.io/), [XCube](https://xcube.readthedocs.io/), MetPy, Holoviz,  etc. All have an Xarray entry point but some particularities makes the workflow still experimental and hacky.

If there was a way to create a massive Zarr/Icechunk/TileDB/Delta Lake on top of a GPFS/Ceph filesystem mounted on a HPC where perhaps access to the data can be accelerated through extra CPU/GPUs that would allow me and my colleagues to build with confidence tooling for some proper Science 😊.

**Comment by martindurant** [comment]
@itsgifnotjiff , that sounds like it might be the perfect opportunity for me to build a vzarr reader rather than a kerchunk indexer (unless @TomNicholas , you don't think it's worth making a difference between the two). I am in Toronto, so if you are two, I'd be happy to meet in person to go over the format and requirements.

**Comment by itsgifnotjiff** [comment]
I am actually in Montreal. 😶 I am however available and motivated. Let me know if you have time to meet virtually if nothing else just for me and my boss to say thank you for your wonderful work. 

[My email for work](yordan.radev@ec.gc.ca)

**Comment by TomNicholas** [comment]
> binary files containing metadata, grid and fortran arrays.

Yet another self-describing binary file format! Sounds like something that could be virtualized.

> We would love to use something like VirtualiZarr to create per model data cubes that can be easily accessed by ML modules like Jax/torch/Dask

Yes, it should be possible to provide access via the zarr-python API.

> also by Web APIs such as pygeoapi for OGC APIs.

I'm less familiar with this, but there might be a way to do that.

> Some of the difficulties are the fact that we have multiple grids and vertical dimensions per file and not all files of a model contain the same variables (accumulation variables start after 0 forecast hour).

The (current) set of restrictions for the virtual zarr approach are [documented here](https://github.com/zarr-developers/VirtualiZarr/pull/532). As your data apparently can already be mapped to the xarray/netCDF data model, as long as you don't have any of those listed issues then I would expect this to be possible.

> data lake

A cloud data lake?

> If there was a way to create a massive Zarr/Icechunk/TileDB/Delta Lake on top of a GPFS/Ceph filesystem mounted on a HPC

This is technically possible using VirtualiZarr and Icechunk together, though currently the support within icechunk for Ceph & and HPC are underdeveloped compared to the typical AWS use case.

---

> perfect opportunity for me to build a vzarr reader rather than a kerchunk indexer

@martindurant it would be great for you to have a go at building a virtualizarr reader (bearing in mind we're just finishing up a [refactor](https://github.com/zarr-developers/VirtualiZarr/issues/498) that will simply the definition of what a virtualizarr reader actually is), but @itsgifnotjiff as the owner of a bespoke file format you should also be the owner of any code designed to read that format. VirtualiZarr is designed to be extensible in that respect, so you can write a dedicated virtualizarr reader for RPN/FST files that lives outside of this main virtualizarr repository.

**Comment by itsgifnotjiff** [comment]
> A cloud data lake?

Internal cloud ... ish 🙂 (I can get into details if needed)

> VirtualiZarr is designed to be extensible in that respect, so you can write a dedicated virtualizarr reader for RPN/FST files that lives outside of this main virtualizarr repository.

That's great to hear! We have published [some tools](https://anaconda.org/fortiers) that allow for advanced workflows such as 

1. Open an RPN file by creating a Polars or Pandas dataframe where each row has the data for a single record and columns for all its metadata and a final column with the actual data (numpy array at the moment).
2. Filter, add, remove etc. in memory as well as perform post proccessing tasks (important for atmospheric sciences) and a new library coming soon that should be able to reproject (horizontally and vertically) any RPN but also Xarray Dataset in memory from any grid/vertical to any grid/vertical called GeoRef (think the completion of the xesmf project by ECMWF).
3. Opening any RPN file/files directly into an Xarray Dataset by using `engine="fstd"` or through [fstd2nc](https://goc-dx.science.gc.ca/~min000/fstd2nc/latest/fstd2nc.html#Transforming-to-xarray).

**Comment by TomNicholas** [comment]
Cool! Let us know how you get on with a virtualizarr reader, and please raise new issues if you have any problems.

**Comment by TomNicholas** [comment]
Oxbow (see this article https://lifeinbytes.substack.com/p/breaking-out-of-bioinformatic-data-silos) is extremely relevant here @nvictus @ianhi

**Comment by nvictus** [comment]
Sorry for chiming in so late. [Oxbow](https://oxbow.readthedocs.io/) currently provides relational Apache Arrow translations for 11 genomic formats. It can do column projection and exploit index files for genomic-range based querying, and in some cases, chunked access. The mismatch to the Zarr/NDArray model is that most of these file formats are heterogeneous in type.

In the case of VCF and BCF, oxbow can retrieve sample data by genotype field as homogeneously-typed tables. But @tomwhite is right that VCF is inherently inefficient. FWIW, oxbow can be a conduit for the conversion from VCF to Zarr or other structures (via arrow). See https://oxbow.readthedocs.io/en/latest/getting-started/quickstart.html

(P.S. I'll be presenting it at SciPy next week!)

**Comment by jhamman** [comment]
Adding the OM-file format (https://github.com/open-meteo/om-file-format) along with these two relevant issues:

- https://github.com/open-meteo/om-file-format/issues/24
- https://github.com/fsspec/kerchunk/issues/464

cc @alxmrs, @patrick-zippenfenig, and @terraputix

**Comment by TomNicholas** [comment]
"Whole-Slide Images" in bio are really just TIFF / JPEG

https://github.com/PABannier/WSIStreamer/issues/33

**Comment by utas-amberbian** [comment]
Adding a real-world example from ocean observing:

I'm a software engineer at AODN (Australia's national ocean data network / IMOS). Our archive is NetCDF on S3 — moorings time series, satellite SST, glider and Argo profiles — served to users via a THREDDS catalogue (https://thredds.aodn.org.au/) plus direct S3 access.

Classic case of a large archival NetCDF collection with growing demand for cloud-native access, without rewriting the archive. I'm interested in whether VirtualiZarr could fit archives like ours — our data is all publicly accessible, so it may be a useful source of real-world test cases.

**Comment by TomNicholas** [comment]
Hi @utas-amberbian .

Your data is in NetCDF format so that part shouldn't be a problem.

> moorings time series, satellite SST, glider and Argo profiles

This will be the hard part - none of this data is particularly neat and "datacube-like". I expect you probably have various ragged dimensions and so on. Dealing with these irregularities should still be doable, but grouping disparate NetCDF files into a single Zarr store will be the hard part.

> a THREDDS catalogue

Whilst you can create HTTP virtual references pointing at the THREDDS server, I wouldn't recommend that - people will hammer your poor server.

> plus direct S3 access

Use the files in S3 instead.

I would be happy to discuss this further - but could you raise a separate issue for that please?

**Comment by utas-amberbian** [comment]
Thanks, Tom — that matches our experience exactly. I've checked with colleagues: the ragged collections (moorings, gliders) we currently handle via Parquet, but for our satellite datacube collections we're interested in virtualising the NetCDF archive to cheaply serve alternate chunking use cases. I'll dig into the file details and open a separate issue with specifics.

**Comment by TomNicholas** [comment]
Presumably every format the the ["NetCDF Expansion Pack"](https://github.com/Intelligent-Data-Design-Inc/NEP) supports reading could also be virtualized as Zarr. So far includes some we haven't tried yet, such as NASA CDF, PDS4, DICOM, and PDB.
